### PR TITLE
Add automatic database detection and binary control

### PR DESCRIPTION
**Pull request by DomBlack**
## Automatic Database Detection

This set of commits adds the ability for the plugin to detect the Encore database and give the user the ability to automatically add it into their IDE.

image
image
image

## Settings Panel

It also adds a new settings panel, which allows the user to disable the plugin, and change which encore binary is being used by the plugin. (This is useful when developing new features)

image

The plugin verifies the binary correctl responds to a call to `encore version` and prevents an invalid binary from being selected:

image

### CODE REVIEW

File: src/main/kotlin/dev/encore/intellij/StartupActivity.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        // Skip our startup if it's not an Encore app
+        if (!isInEncoreApp(project)) {
+            return
+        }
+
+        Encore.LOG.info("Performing startup activity for Encore...")
+
+        // Verify Encore is installed
+        val versionOutput = runEncoreCommand(project, "version") ?: return
+        if (!versionOutput.startsWith("encore version ")) {
```
Comment (eandre) [inline src/main/kotlin/dev/encore/intellij/StartupActivity.kt:25]: Maybe better to just check for zero exit code? Or we should document this requirement in the CLI docs so we don't end up changing this without realizing the effects

File: src/main/kotlin/dev/encore/intellij/settings/SettingsState.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import com.intellij.openapi.components.Storage
+import com.intellij.util.xmlb.XmlSerializerUtil
+
+@State(
+    name = "SettingsState",
+    storages = [Storage("EncorePlugin.xml")]
+)
+class SettingsState : PersistentStateComponent<SettingsState> {
+    var enabled: Boolean = true
+    var encoreBinary: String = "encore"
```
Comment (DomBlack) [inline src/main/kotlin/dev/encore/intellij/settings/SettingsState.kt:15]: These are the original defaults for any settings fields

File: src/main/kotlin/dev/encore/intellij/StartupActivity.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            return
+        }
+        Encore.encoreInstalled = true
+
+        // Get the version
+        val versionString = versionOutput.trim().lines()[0].removePrefix("encore version ")
+        if (versionString.startsWith("v")) {
```
Comment (eandre) [inline src/main/kotlin/dev/encore/intellij/StartupActivity.kt:32]: Maybe we need `encore version --json` or similar :) but good enough for now

**Review by eandre — COMMENTED** [review]
Awesome!

File: src/main/kotlin/dev/encore/intellij/StartupActivity.kt (hunk 1 above)
Comment (DomBlack) [reply src/main/kotlin/dev/encore/intellij/StartupActivity.kt:25]: runEncoreCommand already checks the exit code and returns null if there's a bad exit; but I want to verify that the binary selected is;

1. Encore
2. containing a version number we can parse out for SemVer checks

File: src/main/kotlin/dev/encore/intellij/StartupActivity.kt (hunk 3 above)
Comment (DomBlack) [reply src/main/kotlin/dev/encore/intellij/StartupActivity.kt:32]: In the future yeah :)

**Review by eandre — APPROVED** [review]

**Review by eandre — APPROVED** [review]

**Pull request merged**